drivers/platform/nintendo3ds/mcu/accel.c

```diff
@@ -28,127 +29,146 @@
 #include <linux/iio/driver.h>
 #include <linux/iio/machine.h>
 
-#define REGISTER_ACCEL_MODE	0x40
-#define REGISTER_ACCEL_BASE	0x45
-// XL, XH, YL, YH, ZL, ZH
+#define ACCELEROMETER_OFF	(0)
+#define ACCELEROMETER_ON	(BIT(0))
 
-#define ACCEL_MODE_DISABLE	(0)
-#define ACCEL_MODE_ENABLE	BIT(0)
+#define REG_MODE	0x00
+#define REG_DATA	0x05
+
+#define CTR_ACCEL_NSCALE	(598755)
 
 struct ctr_accel {
-	struct device *dev;
 	struct regmap *map;
 
-	unsigned pwr;
+	int pwr;
+	s16 data[3];
+	unsigned io_addr;
+
 	struct mutex lock;
 };
 
-static int ctr_mcu_accel_toggle_power(struct regmap *map, bool enable)
+static int ctr_accel_set_power(struct ctr_accel *acc, int pwr)
 {
-	return regmap_write(map, REGISTER_ACCEL_MODE,
-		enable ? ACCEL_MODE_ENABLE : ACCEL_MODE_DISABLE);
+	int err = regmap_write(acc->map, acc->io_addr + REG_MODE,
+		(pwr > 0) ? ACCELEROMETER_ON : ACCELEROMETER_OFF);
+
+	if (!err) {
+		acc->pwr = pwr;
+		usleep_range(250, 350);
+		memset(acc->data, 0, sizeof(acc->data));
+	}
+	return err;
 }
 
-static int ctr_mcu_accel_read(struct regmap *map, unsigned reg, int *res)
+static void ctr_accel_update_data(struct ctr_accel *acc)
 {
 	int err;
-	u8 data[2];
-	s16 data_word;
-
-	err = regmap_raw_read(map, REGISTER_ACCEL_BASE + reg, data, 2);
-	if (err)
-		return err;
 
-	data_word = (data[1] << 8) | data[0];
-	*res = data_word;
-	return 0;
+	if (acc->pwr) {
+		err = regmap_bulk_read(acc->map, acc->io_addr + REG_DATA,
+			acc->data, sizeof(acc->data));
+		if (err)
+			memset(acc->data, 0, sizeof(acc->data));
+	}
 }
 
-static int ctr_mcu_accel_read_raw(struct iio_dev *indio_dev,
-								struct iio_chan_spec const *chan,
-								int *val, int *val2, long mask)
+static int ctr_accel_read_raw(struct iio_dev *indio_dev,
+							struct iio_chan_spec const *chan,
+							int *val, int *val2, long mask)
 {
 	int err;
-	struct ctr_accel *accel = iio_priv(indio_dev);
+	struct ctr_accel *acc = iio_priv(indio_dev);
 
+	mutex_lock(&acc->lock);
```

Comment:
what is this mutex protecting? sink into `case II_CHAN_INFO_RAW`?

Reply:
and to make sure acc->pwr is the correct value, hoping there's no other thread that has changed the power state of the accelerometer